docs/tutorials/plugin-provider.md

```diff
@@ -0,0 +1,44 @@
+# Plugin provider
+
+The "Plugin" provider allows to integrate ExternalDNS with DNS providers via an HTTP interface.
+The Plugin provider implements the Provider interface. Instead of implementing code specific to a provider, it implements an HTTP client that sends request to an HTTP API.
+The idea behind it is that providers can be implemented in a separate program: these programs expose an HTTP API that the Plugin provider can interact with. The ideal setup for providers is to run as sidecars in the same pod of the ExternalDNS container and listen on localhost only. This is not strictly a requirement, but we would not recommend other setups.
+
+## Architectural diagram
+
+![Plugin provider](../img/plugin-provider.png)
+
+## API guarantees
+
+Providers implementing the HTTP API have to keep in sync with changes to the Go types `plan.Changes` and `endpoint.Endpoint`. We do not expect to make significant changes to those types given the maturity of the project, but we can't exclude that changes will need to happen. We commit to publishing changes to those in the release notes, to ensure that providers implementing the API can keep providers up to date quickly.
+
+## Implementation requirements
+
+The following table represents the methods to implement mapped to their HTTP method and route.
+
+| Provider method | HTTP Method | Route |
+| --- | --- | --- |
+| Records | GET | /records |
+| ApplyChanges | POST | /records |
+| PropertyValuesEqual | POST | /propertyvaluesequal |
+| AdjustEndpoints | POST | /adjustendpoints |
```

Comment:
I didn't find an API specification for this HTTP API – I guess what you have here is it (plus the definition of the Go interface with the data structures, assuming standard serialization)?

I didn't find an actual documentation of these methods (i.e. what they are supposed to do), but from what I could gather, it's something like this:

* Records – returns all the records which are currently defined.
* ApplyChanges – sends a list of changes to the Records to be applied.
* PropertyValuesEqual – updates a single property. I guess that's for configuration changes?
* AdjustEndpoints – ??? (takes a list of endpoints, returns a similar list)

I'd really recommend writing this down somewhere (at least as comments in provider.go).

From this, I think ApplyChanges should not become a POST, but a PATCH request to /records, as it contains a list of changes to the existing /records resource.
PropertyValuesEqual sounds like it also should be a PATCH to some configuration resource (maybe `PATCH /properties`).
I have no clue what `AdjustEndpoints` is supposed to do, so I can't really recommend something here.

Reply:
To implement a provider, you need to understand the `Provider` interface. This goes regardless of whether you're implementing it embedded or implementing the HTTP protocol.

That the documentation of the `Provider` interface is incomplete is a separate issue. For now you need to read the code.

The ApplyChanges call isn't atomic, so PATCH isn't appropriate. I think the current PR is correct in representing it as a remote procedure call.